syntax = "proto3";

package proxy.filters.http.static_downgrade.v2;

option java_package = "proxy.filters.http.static_downgrade.v2";
option java_outer_classname = "StaticDowngradeProto";
option java_multiple_files = true;

import "api/proxy/common/matcher/v3/matcher.proto";
import "envoy/config/core/v3/base.proto";
import "envoy/type/matcher/v3/string.proto";

import "google/protobuf/wrappers.proto";

import "validate/validate.proto";

message ProtoCommonConfig {
}

message StaticResponse {
  uint32 http_status = 1 [(validate.rules).uint32 = {gte: 200, lt: 600}];
  repeated envoy.config.core.v3.HeaderValue headers = 2;
  envoy.config.core.v3.DataSource body = 3;
}

message DowngradeConfig {
  proxy.common.matcher.v3.CommonMatcher downgrade_rqx = 1;
  proxy.common.matcher.v3.CommonMatcher downgrade_rpx = 2
      [(validate.rules).message.required = true];

  // Decide whether to downgrade according to response flag and response code detail.
  // Response flag should be comma separated flag short string.
  google.protobuf.StringValue response_flags = 4;
  envoy.type.matcher.v3.StringMatcher response_code_details = 5;

  StaticResponse static_response = 3 [(validate.rules).message.required = true];
}

message ProtoRouteConfig {
  // to determine downgrade or not
  proxy.common.matcher.v3.CommonMatcher downgrade_rqx = 1;
  proxy.common.matcher.v3.CommonMatcher downgrade_rpx = 2
      [(validate.rules).message.required = true];

  // Decide whether to downgrade according to response flag and response code detail.
  // Response flag should be comma separated flag short string.
  google.protobuf.StringValue response_flags = 5;
  envoy.type.matcher.v3.StringMatcher response_code_details = 6;

  StaticResponse static_response = 3 [(validate.rules).message.required = true];

  repeated DowngradeConfig specific_configs = 4 [(validate.rules).repeated .max_items = 9];
}
